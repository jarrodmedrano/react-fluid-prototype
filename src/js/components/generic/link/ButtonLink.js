import React from 'react';
import propsAreValid, {externalNavigate, internalNavigate, cleanHtml} from '../../../lib/util';
import {linkPropTypes} from '../../../../data/dataProps';
import { Link } from 'react-router';

class ButtonLink extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            internal: false,
            mosaic: false
        }
    }

    componentWillMount() {
        if (this._isInternal(this.props.to) && this.props.blockType !== "buttonExternal") {
            this.setState({
                internal: true
            })
        }

        if (this.props.layout && this.props.layout === 'mosaic') {
            this.setState({
                mosaic: true
            })
        }
    }

    _isInternal(to) {
        // If it's a relative url such as '/path', 'path' and does not contain a protocol we can assume it is internal.
        if (to.startsWith('/')) return true;
    }

    _handleClick(target, internal) {
        if (internal) {
            internalNavigate(target, this.context.router);
        } else {
            externalNavigate(target);
        }
    }

    _buildLink() {
        const {to, children, ...rest} = this.props;

        let linkProps = {
            onClick: (e) => {
                e.preventDefault();
                this._handleClick(to, this.state.internal)
            },
            draggable: 'false',
        };

        const linkElement = React.createElement(
            Link,
            {
                ...linkProps,
                to: to,
                href: to,
                ...rest,
                dangerouslySetInnerHTML: {__html: cleanHtml(children)}
            }
        );

        const mosaicLinkElement = React.createElement(
            Link,
            {
                ...linkProps,
                to: to,
                href: to,
                className: 'mosaic-link'
            },
            children
        );

        if(this.state.mosaic === true) {
            return mosaicLinkElement
        } else {
            return linkElement
        }
    }

    render() {
        if (propsAreValid(this.props, this)) {
            return this._buildLink();
        }
        return null
    }
}

ButtonLink.defaultProps = {
    to: '',
    children: ''
};

ButtonLink.propTypes = linkPropTypes;

ButtonLink.contextTypes = {
    router: React.PropTypes.object,
};

export default ButtonLink;
