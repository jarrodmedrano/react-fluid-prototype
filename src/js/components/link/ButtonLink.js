import React from 'react';
import classNames from 'classnames';
import './button-link.scss!';
import propsAreValid, {externalNavigate} from '../../lib/util';
import {linkPropTypes} from '../../../data/dataProps';
import {Link} from '../../lib/scroll';
import sanitizeHtml from 'sanitize-html';

class ButtonLink extends React.Component {

    _parseTo(to) {
        let parser = document.createElement('a');
        parser.href = to;
        return parser;
    }

    _isInternal(to) {
        // If it's a relative url such as '/path', 'path' and does not contain a protocol we can assume it is internal.
        if(to.indexOf("://")=== -1 && to.indexOf("ms-")=== -1) return true;
        const toLocation = this._parseTo(to);
        return window.location.hostname === toLocation.hostname;
    }

    cleanHtml(dirty) {
        return sanitizeHtml(dirty, {
            allowedTags: ['b', 'i', 'em', 'strong', 'a', 'span'],
            allowedAttributes: {
                'a': ['href', 'style'],
                'span': ['style'],
                'b': ['style'],
                'p': ['style']
            }
        });
    }

    _handleClick(target) {
        externalNavigate(target);
    }

    render() {
        if(propsAreValid(this.props, this)) {
            const {to, children, layout, ...rest} = this.props;
            const _isInternal = this._isInternal(to);
            let templateClass = classNames('c-action-trigger');

            if (_isInternal) {
                if(layout === 'mosaic') {
                    return (
                        <Link className={templateClass} activeClass="active" to={to} {...rest} spy={true} smooth={true} duration={500} isDynamic={true} draggable="false" />
                    )
                } else {
                    return (
                        <Link className={templateClass} activeClass="active" to={to} {...rest} spy={true} smooth={true} duration={500} isDynamic={true} dangerouslySetInnerHTML={{ __html: this.cleanHtml(children) }} draggable="false" />
                    );
                }
            } else {
                if(layout === 'mosaic') {
                    return (<a href={to} {...rest} onClick={(e) => {e.preventDefault(); this._handleClick(to)}} draggable="false">{children}</a>);
                } else {
                    return (<a href={to} {...rest} onClick={(e) => {e.preventDefault(); this._handleClick(to)}} dangerouslySetInnerHTML={{ __html: this.cleanHtml(children) }} draggable="false" />);
                }
            }
        } return null
    }
}

ButtonLink.defaultProps = {
    to: '',
    children: ''
};

ButtonLink.propTypes = linkPropTypes;

export default ButtonLink;