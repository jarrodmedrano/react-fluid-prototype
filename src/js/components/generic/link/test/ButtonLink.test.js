import React from 'react';
import renderer from 'react-test-renderer';
import ButtonLink from '../ButtonLink';
import {shallow} from 'enzyme';

describe('Button Link functions', () => {

    const minProps = {
        dangerouslySetInnerHTML: {},
        draggable: "false",
        href: "",
        to: ""
    };

    it('should click', () => {
        const spy = jasmine.createSpy();
        const wrapper = shallow(<ButtonLink onClick={spy} {...minProps} />);
        wrapper.find('a').simulate('click');
        expect(spy).toHaveBeenCalled();
        expect(wrapper).toMatchSnapshot();
    });

    it('Should have internal state false when linking externally', () => {
        const wrapper = shallow(<ButtonLink to="http://www.google.com" />);
        expect(wrapper.state('internal')).toEqual(false);
        expect(wrapper).toMatchSnapshot();
    });

    it('Should have internal state false when link has a token', () => {
        const wrapper = shallow(<ButtonLink to="media:{pictureslibrary}//Glaciers.docx" />);
        expect(wrapper.state('internal')).toEqual(false);
        expect(wrapper).toMatchSnapshot();
    });

    it('Should have internal state true when linking internally', () => {
        const wrapper = shallow(<ButtonLink to="/windows/#Cortana" />);
        expect(wrapper.state('internal')).toEqual(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render without exploding', () => {
        const component = renderer.create(
            <ButtonLink {...minProps} />
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
