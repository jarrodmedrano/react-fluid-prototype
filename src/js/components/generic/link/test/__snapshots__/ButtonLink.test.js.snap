// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Link functions Should have internal state false when link has a token 1`] = `
<a
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  draggable="false"
  href="media:{pictureslibrary}//Glaciers.docx"
  onClick={[Function]}
  to="media:{pictureslibrary}//Glaciers.docx"
/>
`;

exports[`Button Link functions Should have internal state false when linking externally 1`] = `
<a
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  draggable="false"
  href="http://www.google.com"
  onClick={[Function]}
  to="http://www.google.com"
/>
`;

exports[`Button Link functions Should have internal state true when linking internally 1`] = `
<a
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  draggable="false"
  href="/windows/#Cortana"
  onClick={[Function]}
  to="/windows/#Cortana"
/>
`;

exports[`Button Link functions should click 1`] = `
<a
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  draggable="false"
  href=""
  onClick={[Function]}
  to=""
/>
`;

exports[`Button Link functions should render without exploding 1`] = `
<a
  dangerouslySetInnerHTML={
    Object {
      "__html": "",
    }
  }
  draggable="false"
  href=""
  onClick={[Function]}
  to=""
/>
`;
